### PR TITLE
Add MCP template sync tools (status + sync) for #21

### LINKED ISSUE #21: Add tools to assist in working with extension
Knowing if a template is linked, doing a sync operation, etc

### PR DESCRIPTION
**Pull request by totallynotjon**
## What

Closes #21 (first slice). Exposes the extension's local **file ↔ Rewst-template sync** over MCP so an AI assistant can see a file's link/sync state and push or pull a linked template — without driving the VS Code UI. Two new MCP-only tools:

| Tool | Access | Gating | Purpose |
|---|---|---|---|
| `buddy_template_sync_status` | read | ungated (derives org from the link) | Is this file linked? How does it compare to remote — `in-sync` / `remote-only` / `local-ahead` / `conflict` — plus orgId, templateId, sync-on-save, recommended direction. Changes nothing. |
| `buddy_template_sync` | write | write gate + `writeOrgAllowlist` (whole tool); per-call approval **on upload only** | Push/pull a linked file. `direction` defaults to `auto`. |

### Why the new tools (not the existing sync command)

The existing save-driven sync resolves conflicts with an **interactive VS Code modal**, which an external MCP client can never answer. So `buddy_template_sync` replaces the modal with an explicit `direction`:

- **`auto`** (default): uploads when only local changed, downloads when local is empty, refreshes metadata when bodies match, and on a **conflict changes nothing** — it returns a `conflict` status and asks the caller to choose. Never a surprise overwrite.
- **`upload`** overwrites Rewst with the local file; **`download`** overwrites the local file with Rewst.

### Safety

- `buddy_template_sync` is a single `access:'write'` capability, so the central `McpActions` gate requires write tools enabled + an allowlisted org for **every** direction (fails closed); only **upload** additionally prompts for approval and changes Rewst.
- Cross-org by-path reach is blocked: the handler verifies `link.org.id === orgId` **and** `remoteTemplate.orgId === orgId` before any upload.
- An ambiguous bare filename that matches two linked files resolves to nothing (refused) rather than picking one.
- An explicit `upload` of an empty local file (which clears the remote body) gets an explicit warning in the approval prompt and the result.

### Implementation

- `SyncManager` refactored to share the non-interactive decision (`computeSyncDecision`) and metadata refresh (`refreshLinkMetadata`) with the existing on-save path — both decide identically; the modal stays confined to the on-save conflict branch. Behavior-preserving for save-on-sync.
- New `templateSyncCapabilities.ts`; reuses the shared `mutationApproval.ts` helpers (no copied approval plumbing).

## Tests

37 unit tests in `templateSyncCapabilities.test.ts` (every direction × action, approval grant/deny/reuse, conflict no-op, org guards, descriptor flags, `matchLinkByPath` + live-`LinkManager` `resolveLinkedUri` resolution incl. ambiguity/normalization). Full unit suite green (689 passing); typecheck, eslint, markdownlint, and `changelog:check` all clean.

## Review

Authored and ran an adversarial multi-agent review (5 dimensions → independent per-finding verification). All 10 confirmed findings are addressed in this PR (gating-text parity, shared-helper reuse, empty-upload warning, ambiguous-path refusal, error-context forwarding, and the added descriptor/`resolveLinkedUri`/edge-case tests).

## Deferred to follow-ups (issue #21 remainder)

- `buddy_template_link` / `buddy_template_unlink` (link a template id ↔ a file URI) and a sync-on-save toggle.
- Deep-copy of a nested/bundle into new templates.
- Optional: split status/download into a read-tier path so download works with write tools off (kept as one write tool here for ergonomics); a token-gated integration test that drives the real `defaultTemplateSyncDeps` end-to-end.

## Summary by CodeRabbit

* **New Features**
  * Added MCP-based template sync capabilities to check sync status and synchronize linked template files (auto/upload/download), including metadata refresh and conflict handling.
  * Added MCP tools to link/unlink templates and toggle sync-on-save for linked files.
  * Added MCP capability to deep-clone a template bundle across linked templates with approval and rollback safeguards.
* **Tests**
  * Added/expanded unit test coverage for sync, linking, sync-on-save, and bundle cloning behavior, metadata, validation, and capability descriptors.

### CODE REVIEW

File: src/capabilities/templateSyncCapabilities.ts [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+	// A session can manage several orgs, so confirm the linked file AND the remote
+	// template belong to the requested org before any upload — otherwise a file
+	// linked to a sibling org could be reached by path.
+	if (link.org.id !== orgId) {
+		throw new Error(`"${uri}" is linked to org ${link.org.id}, not ${orgId}.`);
+	}
+	const remoteOrgId = (remoteTemplate as { orgId?: unknown }).orgId;
+	if (typeof remoteOrgId === 'string' && remoteOrgId !== orgId) {
+		throw new Error(`Template ${link.template.id} is not in org ${orgId}.`);
+	}
```
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateSyncCapabilities.ts:277]: Good catch — fixed in 54d8a6c. The remote-org guard now fails closed: it rejects when `remoteOrgId` is not a string as well as on a mismatch (`typeof remoteOrgId !== 'string' || remoteOrgId !== orgId`), so the org re-verification can never be bypassed before a mutation. Added a `fails closed when the remote template has no orgId` unit test.

File: src/capabilities/templateSyncCapabilities.ts [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+
+	if (effective === 'metadata') {
+		await deps.refreshMetadata(target);
+		return JSON.stringify(
+			{
+				status: 'in-sync',
+				templateId: link.template.id,
+				name: link.template.name,
+				message: 'Bodies already match; refreshed the local link metadata. Nothing was uploaded or downloaded.',
+			},
```
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateSyncCapabilities.ts:305]: Fixed in 54d8a6c. The metadata-sync (`in-sync`) response now returns `remoteTemplate.name` (the refreshed value) instead of the stale `link.template.name`, so a remote rename is reflected. Added a `metadata-sync returns the refreshed remote template name` unit test.

File: src/capabilities/inputHelpers.ts [hunk 3]
```diff
@@ -4,6 +4,35 @@
...
+	sessions: readonly Session[],
+	getTemplate: (session: Session, templateId: string) => Promise<FullTemplateFragment>,
+	templateId: string,
+): Promise<{ template: FullTemplateFragment; session: Session } | undefined> {
+	for (const session of sessions) {
+		try {
+			return { template: await getTemplate(session, templateId), session };
+		} catch {
+			// Try the next session.
+		}
```
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/inputHelpers.ts:37]: Fixed in 5a1c513. `getTemplateFromAnySession` now only continues to the next session on a `not found` error; any auth/network/SDK failure is remembered and, if no session resolves the id, re-thrown rather than masked as `template_not_found`. (Re-throwing immediately would abort the legitimate multi-session fallback on a transient error in an earlier session, so it surfaces only when every session failed and at least one failed operationally.) Added a unit test that an operational fetch error propagates instead of reporting `template_not_found`.

File: src/capabilities/templateCloneCapabilities.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+			if (nodes.size >= maxTemplates) {
+				skippedCap.push(refKey);
+				continue;
+			}
+			let referenced: FullTemplateFragment;
+			try {
+				referenced = await deps.getTemplate(sourceSession, refKey);
+			} catch {
+				missingReferences.push(refKey);
+				continue;
```
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateCloneCapabilities.ts:205]: Fixed in 5a1c513. The clone graph traversal now classifies a referenced-template fetch the same way: only a genuine `not found` becomes a `missingReference`; an operational failure (auth/network/SDK) is re-thrown, aborting the walk before any templates are created — so a transient error can no longer yield a partial clone with stale references. Added a unit test asserting no clones are created when a reference fetch fails operationally.

File: src/capabilities/templateCloneCapabilities.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+			},
+			maxDepth: { type: 'number', description: 'Max reference depth to walk. Default 10, capped at 25.' },
+		},
+		required: ['orgId', 'rootTemplateId'],
+	},
+};
+
+export const TEMPLATE_CLONE_CAPABILITIES: Capability[] = [
+	{ spec: bundleCloneSpec, access: 'write', chat: false, mcp: true, run: (input, ctx) => runBundleClone(input, ctx) },
+];
```
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateCloneCapabilities.ts:318]: Done in 5a1c513 — added `requiresOrg: true` explicitly to the `buddy_template_bundle_clone` descriptor. (Behavior was already org-scoped via the truthy default, but making it explicit locks the contract.)

File: src/capabilities/templateCloneCapabilities.test.ts [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+			const ok = createMockSession();
+			ok.wrapper.when('deleteTemplate', { data: { deleteTemplate: 'x' } });
+			await defaultTemplateCloneDeps.deleteTemplate(ok.session, 'x');
+
+			const bad = createMockSession();
+			bad.wrapper.when('deleteTemplate', { data: { deleteTemplate: null } });
+			await assert.rejects(() => defaultTemplateCloneDeps.deleteTemplate(bad.session, 'x'), /rollback delete failed/);
+		});
+	});
+});
```
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateCloneCapabilities.test.ts:392]: Added in 5a1c513: `src/test/integration/templateClone.test.ts`, following the established write-capability pattern (`templateMutate.test.ts` et al.) — opt-in via `REWST_TEST_WRITE=1`, always targets the token's own primary (sandbox) org. It creates a real root+child bundle, runs `buddy_template_bundle_clone` through the live approval + create + updateTemplate flow, asserts the cloned root's body was rewritten to the cloned child id (not the original), and deletes every created template (sources and clones) in `finally`.

File: src/capabilities/inputHelpers.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/inputHelpers.ts:37]: Fixed in 5a1c513.

`getTemplateFromAnySession` now continues to the next session only on a genuine "not found" (the id simply belongs to a different session's org). Any auth/network/SDK failure is remembered and re-thrown if no session resolves the id, so an outage can no longer be reported as a `template_not_found`.

One deliberate nuance versus the suggested patch: rather than re-throwing on the *first* operational error, it defers the throw to the end of the loop. That keeps the multi-session fallback intact — a healthy later session can still resolve the template — and only surfaces the remembered operational error when nothing resolves it.

Covered by a new `Unit: templateLinkCapabilities` test, `propagates an operational fetch error instead of reporting template_not_found`.

File: src/capabilities/templateCloneCapabilities.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateCloneCapabilities.ts:205]: Fixed in 5a1c513.

The traversal `catch` now classifies the failure: only a genuine "not found" is recorded as a `missingReference`; a transient auth/network/SDK error is re-thrown. Traversal (Phase 1) runs before any create/update (Phase 3), so that abort happens before a single template is created — there's no partial clone left pointing at stale references, and nothing to roll back.

Covered by a new `Unit: templateCloneCapabilities` test, `aborts without creating anything when a reference fetch fails operationally`.

File: src/capabilities/templateCloneCapabilities.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateCloneCapabilities.ts:318]: Fixed in 5a1c513. `buddy_template_bundle_clone` now sets `requiresOrg: true` explicitly in `TEMPLATE_CLONE_CAPABILITIES`, so the org-scoped execution semantics are pinned in the descriptor itself instead of relying on the default.

File: src/capabilities/templateCloneCapabilities.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (totallynotjon) [reply src/capabilities/templateCloneCapabilities.test.ts:392]: Added in 5a1c513: `src/test/integration/templateClone.test.ts`, following the `templateMutate.test.ts` pattern — opt-in (`REWST_TEST_WRITE=1` + a token) and sandbox-only (it targets the token's own primary org).

It creates a real root template that references a child, deep-clones the bundle through the capability — exercising the end-to-end approval → create → `updateTemplate` flow against a live session — and asserts the cloned root's body was rewritten to the cloned child's new id (and the original child id is gone). Every template it creates, sources and clones, is deleted in a `finally` block even if an assertion fails.
